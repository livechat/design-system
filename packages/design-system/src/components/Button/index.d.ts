import * as React from 'react';

export type ButtonSize = 'compact' | 'large';

export interface ButtonProps {
  children?: string;
  id?: string;
  primary?: boolean;
  destructive?: boolean;
  disabled?: boolean;
  loading?: boolean;
  size?: ButtonSize;
  outline?: boolean;
  fullWidth?: boolean;
  icon?: any;
  submit?: boolean;
  accessibilityLabel?: string;
  ariaControls?: string;
  ariaExpanded?: boolean;
  onClick?(): void;
  onFocus?(): void;
  onBlur?(): void;
  renderIcon?: string | React.ReactNode;
}
