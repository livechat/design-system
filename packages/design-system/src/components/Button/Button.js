import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './style.css';

const cx = classNames.bind(styles);
const acceptedSizes = ['large', 'compact'];

const Button = props => {
  const {
    children,
    primary,
    destructive,
    disabled,
    loading,
    size,
    fullWidth,
    submit,
    accessibilityLabel,
    ariaControls,
    ariaExpanded,
    ...buttonProps
  } = props;

  const isDisabled = disabled || loading;
  const type = submit ? 'submit' : 'button';
  let buttonType = null;

  if (primary) {
    buttonType = 'primary';
  } else if (destructive) {
    buttonType = 'destructive';
  }

  return (
    <button
      {...buttonProps}
      type={type}
      disabled={isDisabled}
      role={loading ? 'alert' : undefined}
      aria-busy={loading ? true : undefined}
      aria-label={accessibilityLabel}
      aria-controls={ariaControls}
      aria-expanded={ariaExpanded}
      className={cx({
        btn: true,
        'btn--full-width': fullWidth,
        'btn--primary': buttonType === 'primary',
        'btn--destructive': buttonType === 'destructive',
        [`btn--${size}`]: acceptedSizes.some(s => s === size)
      })}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.node.isRequired,
  id: PropTypes.string,
  /**
   * Type of button
   */
  primary: PropTypes.bool,
  /**
   * Type of button
   */
  destructive: PropTypes.bool,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  /**
   * Size of button
   */
  size: PropTypes.oneOf(['compact', 'large']),
  /**
   * Sets button width to max-width=320px
   */
  fullWidth: PropTypes.bool,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  submit: PropTypes.bool,
  className: PropTypes.string,
  accessibilityLabel: PropTypes.string,
  ariaControls: PropTypes.string,
  ariaExpanded: PropTypes.string
};

export default Button;
