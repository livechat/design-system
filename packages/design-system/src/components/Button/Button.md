### Primary Button

Use when there is an action that is more important than any other actions and you need to draw attention to it. Don’t use more than one primary button per screen to avoid confusing users. Not every screen needs a primary button.
```js
<Button primary>Primary Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button primary>Primary Button</Button>
  </ComponentHtmlMarkup>
```
### Basic Button

The standard button for most use cases. Only use another style if a button requires more or less visual attention.
```js
  <Button>Basic Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button>Basic Button</Button>
  </ComponentHtmlMarkup>
```
### Destructive Button

Use as a final confirmation for a destructive action such as deleting. These are found mostly in confirmation modals.
```js
  <Button destructive>Destructive Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button destructive>Destructive Button</Button>
  </ComponentHtmlMarkup>
```
### Button Sizes
#### Compact
Use in a table or list where space is tight.
```js
  <Button size="compact" primary>Compact Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button size="compact" primary>Compact Button</Button>
  </ComponentHtmlMarkup>
```
#### Regular
```js
  <Button primary>Regular Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button primary>Regular Button</Button>
  </ComponentHtmlMarkup>
```
#### Large
Use for the main call to action in empty states or for calls to action shown with large illustrations.
```js
<Button size="large" primary>Large Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button size="large" primary>Large Button</Button>
  </ComponentHtmlMarkup>
```
#### Full-width

Button expanded to the full width to fill the parent container (max-width: 320px)
```js
  <Button fullWidth primary>Full-width Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button fullWidth primary>Full-width Button</Button>
  </ComponentHtmlMarkup>
```

### Best practices
 - Prioritize most important action; don’t use more than 1 primary button per screen.
 - Position buttons in consistent locations in the UI.
 - Use short yet descriptive labels ({verb} + {noun}).
 - Use strong verbs to encourage action.
 - Avoid unnecessary words and articles such as the, an, or a.