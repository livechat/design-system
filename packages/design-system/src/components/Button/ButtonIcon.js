import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import getMergedClassNames from '../../utils/getMergedClassNames';
import styles from './style.scss';

const cx = classNames.bind(styles);

export const ButtonIcon = ({ position, children }) => {
  const baseClass = 'btn__icon';
  const classes = getMergedClassNames(
    cx({
      [`${baseClass}`]: true,
      [`${baseClass}-left`]: position === 'left',
      [`${baseClass}-right`]: position === 'right'
    })
  );
  return <i className={classes}>{children}</i>;
};

ButtonIcon.propTypes = {
  position: PropTypes.oneOf(['left', 'right']).isRequired,
  children: PropTypes.node.isRequired
};
