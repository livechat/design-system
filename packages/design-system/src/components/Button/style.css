.btn {
  max-width: 320px;
  position: relative;
  align-items: center;
  justify-content: center;
  user-select: none;
  text-decoration: none;
  transition-property: opacity, border, box-shadow;
  transition-duration: 200ms;
  transition-timing-function: cubic-bezier(0.64, 0, 0.35, 1);
  border-width: 1px;
  border-style: solid;
  text-align: center;
  cursor: pointer;
  border-radius: 4px;
  font-weight: 600;
  display: inline-flex;
  background-color: #fff;
  color: rgba(66,77,87,1);
  border-color: #bcc6d0;
  font-size: 15px;
  padding: 0 16px;
  min-width: 36px;
  min-height: 36px;
  line-height: 36px;
}

.btn:hover {
  opacity: 0.8;
}

.btn:focus {
  outline: 0;
  box-shadow: 0 0 0 1px #427fe1;
}

.btn:disabled {
  opacity: 0.65;
  cursor: not-allowed;
}

.btn .icon {
  color: #4384f5;
  margin-right: 7px;
}

.btn--full-width {
  display: flex;
  width: 100%;
}

.btn--primary {
  background-color: #4384f5;
  color: #fff;
  border-color: #4384f5;
}

.btn--primary:focus {
  box-shadow: 0 0 0 1px #427fe1;
}

.btn--destructive {
  background-color: #f4574c;
  color: #fff;
  border-color: #f4574c;
}

.btn--destructive:focus {
  box-shadow: 0 0 0 1px #f4574c;
}

.btn--large {
  font-size: 15px;
  padding: 0 24px;
  min-width: 42px;
  min-height: 42px;
  line-height: 42px;
}

.btn--compact {
  font-size: 14px;
  padding: 0 16px;
  min-width: 32px;
  min-height: 32px;
  line-height: 32px;
}
