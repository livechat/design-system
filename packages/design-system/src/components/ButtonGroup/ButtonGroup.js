import * as React from 'react';
import * as PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './style.scss';

const defaultIndex = -1;
const baseClass = 'btn-group';

class ButtonGroup extends React.Component {
  state = {
    index: defaultIndex
  };

  static getDerivedStateFromProps(props, state) {
    return {
      index:
        typeof props.currentIndex === 'number'
          ? props.currentIndex
          : state.currentIndex
    };
  }

  handleClick = (index, event) => {
    this.props.onIndexChange && this.props.onIndexChange(index, event);
    this.setState({ currentIndex: index });
  };

  render() {
    const { index } = this.state;
    const { size, fullWidth, children, className, onIndexChange, currentIndex, ...restProps } = this.props;

    const mappedChildren = React.Children.map(children, (child, i) =>
      React.cloneElement(child, {
        size,
        fullWidth,
        primary: false,
        secondary: true,
        destructive: false,
        type: 'button',
        onClick: event => {
          this.handleClick(i, event);
          if (child.props.onClick) {
            child.props.onClick(event);
          }
        },
        className: cx({
          [className]: !!className,
          [styles.active]: i === index
        })
      })
    );

    return (
      <div role="group" className={styles[baseClass]} {...restProps}>
        {mappedChildren}
      </div>
    );
  }
}

ButtonGroup.propTypes = {
  /**
   * If `true`, the buttons in group will take up the full width of its container
   */
  fullWidth: PropTypes.bool,

  /**
   * Additionall class name for buttons components
   */
  className: PropTypes.string,

  /**
   * Callback fired when the value of `currentIndex` changes
   */
  onIndexChange: PropTypes.func,

  /**
   * The index of button in group which is currently active
   */
  currentIndex: PropTypes.number,

  /**
   * Size of buttons in group
   */
  size: PropTypes.string,

  /**
   * One or more `Button` components
   */
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ])
};

export default ButtonGroup;
