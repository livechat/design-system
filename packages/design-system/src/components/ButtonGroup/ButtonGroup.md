### Basic ButtonGroup

By default ButtonGroup use internal state to track currently selected button. It fires `onIndexChange`callback each time when one of buttons was clicked supling its index.

```js
<ButtonGroup>
  <Button>First Button</Button>
  <Button>Second Button</Button>
  <Button>Third Button</Button>
</ButtonGroup>
```

```js noeditor
<ComponentHtmlMarkup>
  <ButtonGroup>
    <Button>First Button</Button>
    <Button>Second Button</Button>
    <Button>Third Button</Button>
  </ButtonGroup>
</ComponentHtmlMarkup>
```

### Controlled ButtonGroup

If you want to use ButtonGroups as controlled component you can set `currentIndex` props to set which button is active.

```js
initialState = { currentIndex: 0 };

handleClick = (index, e) => {
  setState({
    currentIndex: index
  })
}

<ButtonGroup currentIndex={state.currentIndex} onIndexChange={handleClick}>
  <Button>First Button</Button>
  <Button>Second Button</Button>
  <Button>Third Button</Button>
</ButtonGroup>
```

```js noeditor
<ComponentHtmlMarkup>
  <ButtonGroup currentIndex={1}>
    <Button>First Button</Button>
    <Button>Second Button</Button>
    <Button>Third Button</Button>
  </ButtonGroup>
</ComponentHtmlMarkup>
```
