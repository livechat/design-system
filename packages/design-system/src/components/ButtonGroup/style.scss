@import "../../scss/variables/colors";

.btn-group {
  display: inline-flex;
  flex-direction: row;
  vertical-align: middle;

  .btn {
    &.active {
      color: $font-primary-color;
      border-color: $border-primary-color;
      background-color: $ui-background-color;
      z-index: 1;

      &:hover {
        background-color: $ui-background-color-hover;
      }
    }

    &:not(:first-child) {
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
    }

    &:not(:last-child) {
      border-right: 0;
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    }
  }
}
