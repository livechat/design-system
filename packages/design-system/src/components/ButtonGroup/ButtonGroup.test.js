import * as React from 'react';
import { shallow } from 'enzyme';
import ButtonGroup from './ButtonGroup';
import Button from '../Button';

const onIndexChangeFn = jest.fn();

describe('ButtonGroup', () => {
  it('should render empty button group', () => {
    const component = shallow(<ButtonGroup />);

    expect(component).toMatchSnapshot();
  });

  it('should render button group with button', () => {
    const component = shallow(
      <ButtonGroup>
        <Button />
      </ButtonGroup>
    );

    expect(component).toMatchSnapshot();
  });

  it('should handle current index change', () => {
    const component = shallow(
      <ButtonGroup onIndexChange={onIndexChangeFn}>
        <Button id="button1" />
        <Button id="button2" />
        <Button id="button3" />
      </ButtonGroup>
    );

    const fakeEvent = new Event('fake');

    component.find('#button1').simulate('click', fakeEvent);
    expect(onIndexChangeFn).toBeCalledWith(0, fakeEvent);

    component.find('#button2').simulate('click', fakeEvent);
    expect(onIndexChangeFn).toBeCalledWith(1, fakeEvent);

    component.find('#button3').simulate('click', fakeEvent);
    expect(onIndexChangeFn).toBeCalledWith(2, fakeEvent);
  });

  it('should use current index from porps', () => {
    const component = shallow(
      <ButtonGroup currentIndex={1}>
        <Button />
      </ButtonGroup>
    );

    expect(component.state('currentIndex')).toBe(1);
  });
});
