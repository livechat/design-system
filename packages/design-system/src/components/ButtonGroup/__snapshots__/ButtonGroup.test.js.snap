// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should render button group with button 1`] = `
<div
  className="btn-group"
  onIndexChange={[Function]}
  role="group"
>
  <Component
    className=""
    destructive={false}
    key=".0"
    onClick={[Function]}
    primary={false}
    secondary={true}
  />
</div>
`;

exports[`ButtonGroup should render empty button group 1`] = `
<div
  className="btn-group"
  onIndexChange={[Function]}
  role="group"
/>
`;
